quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -0,0 +1,830 @@
+    match expr {
+        Expression::Address(_)
+        | Expression::Number(_)
+        | Expression::PiConstant
+        | Expression::Variable(_) => 1,
+        Expression::FunctionCall(FunctionCallExpression {
+            function: _,
+            expression,
+        }) => 1 + size(expression),
+        Expression::Infix(InfixExpression {
+            left,
+            operator: _,
+            right,
+        }) => 1 + size(left) + size(right),
+        Expression::Prefix(PrefixExpression {
+            operator: _,
+            expression,
+        }) => 1 + size(expression),
+    }
+}
+
+// It's verbose to go alone! Take this.
+macro_rules! infix {
+    ($left:expr, $op:expr, $right:expr) => {
+        Expression::Infix(InfixExpression {
+            left: $left.into(),
+            operator: $op,
+            right: $right.into(),
+        })
+    };
+}
+macro_rules! add {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Plus, $right)
+    };
+}
+macro_rules! sub {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Minus, $right)
+    };
+}
+macro_rules! mul {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Star, $right)
+    };
+}
+macro_rules! div {
+    ($left:expr, $right:expr) => {
+        infix!($left, InfixOperator::Slash, $right)
+    };
+}
+
+// Are these both of the form something * x for the _same_ x?
+fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
+    match (left_ax, right_ax) {
+        (
+            Expression::Infix(InfixExpression {
+                left: ref ll,
+                operator: InfixOperator::Star,
+                right: ref lr,
+            }),
+            Expression::Infix(InfixExpression {
+                left: ref rl,
+                operator: InfixOperator::Star,
+                right: ref rr,
+            }),
+        ) => **ll == **rl || **ll == **rr || **lr == **rl || **lr == **rr,
+        _ => false,
+    }
+}
+
+fn simplify_infix(l: &Expression, op: InfixOperator, r: &Expression, limit: u64) -> Expression {
+    if limit == 0 {
+        // bail
+        return Expression::Infix(InfixExpression {
+            left: l.clone().into(),
+            operator: op,
+            right: r.clone().into(),
+        });
+    }
+    // There are … many cases here
+    match (
+        simplify(l, limit.saturating_sub(1)),
+        op,
+        simplify(r, limit.saturating_sub(1)),
+    ) {
+        //----------------------------------------------------------------
+        // First: only diving one deep, pattern matching on the operation
+        // (Constant folding and cancellations, mostly)
+        //----------------------------------------------------------------
+
+        // Addition and Subtraction
+
+        // Adding with zero
+        (Expression::Number(x), InfixOperator::Plus, right) if is_zero(x) => right,
+        (left, InfixOperator::Plus, Expression::Number(x)) if is_zero(x) => left,
```

Comment:
Not sure if this improves things:

```suggestion
        (Expression::Number(x), InfixOperator::Plus, other) | (other, InfixOperator::Plus, Expression::Number(x)) if is_zero(x) => other,
```